File: lab4/app/runners.go
package main

import (
	"fmt"
	"time"
	"math/rand"
)

func doRun(area int) {
	for i := 0; i < area; i++ {
		var barrier int = rand.Intn(2);
		if (barrier != 0) {
			amt := time.Duration(rand.Intn(4))
			time.Sleep(time.Millisecond * amt)
			fmt.Println("barrier number: ", i, "\n")
		}
	}
}

func main() {
	hight := rand.Intn(25);
	runners := 3;
	area := hight * runners;

	for i := 0; i < runners; i++ {
        fmt.Println("Runner ", i + 1, "start\n");        
		go doRun(area);
	}

	var input string
	fmt.Scanln(&input)
}

File: lab4/app/runners.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>

void doRun(int area)
{
    int runner = 0;
    for (int i = 1; i < area+1; i++) {
        int barrier = rand() % 2;
        if (barrier == 0) {
            ++runner;
        }
        else if (barrier == 1) {
            sleep(rand() % 4);
            printf("barrier number: %d\n", i);
            ++runner;
        }
    }    
}

int main()
{
    int hight = rand() % 25;
    int runners = 3;
    int area = runners * hight;
    
    for (int i = 0; i < runners; i++) {
        pid_t pid = fork();
        if (pid < 0) {
            perror("Fork failed");
        }
        if (pid == 0) {
            printf("Start runner number: %d\n", (int) getpid());
            doRun(area);
            exit(0);
        }
    }
    wait(NULL);
    
    return 0;
}
